Fix: Prevent ICE on final switch forward referencing its enum

abulgit opened the pull request:
Closes: https://github.com/dlang/dmd/issues/20867
Fix Details-
1. First I Added a check to detect when a `final switch` statement is used within an enum's own definition
2. Another secondary check to detect incomplete enum member values during the missing members check

It now properly detects and reports the circular reference with the error message: "cannot use `final switch` on enum `%s` while it is being defined".

Testing-
I added a testcase to verify the fix: `fail_compilation/fix20867.d`

This test includes:
Test Case 1: The exact scenario from the GitHub issue
Test Case 2: A variation with multiple enum members
Test Case 3: A regular use of final switch (should compile normally)
Test Case 4: A nested circular reference case

All test cases Work Fine.

abulgit committed:
Fix: Prevent ICE on final switch forward referencing its enum

dkorpel commented:
Props for tackling this bug, forward references can be tricky. :+1: 

Your code for inferring whether an enum is incomplete looks functional, but there's a simpler way to check the semantic state of symbols that the rest of the compiler uses: the `semanticRun` variable in `Dsymbol` stores the last completed pass, so this should work better:

```
if (ed.semanticRun < PASS.semanticdone)
    error(...);
```

For the tests, please minimize the amount of added test code, since the test suite is already so large.

2. The presence of multiple enum members doesn't change any code paths
3. Regular final switch is already tested somewhere else
4. The nested case makes sense for testing your code that walks up the scopes, but it will be redundant with the simpler check.

So just keep test 1.

abulgit committed:
more simpler approach

abulgit commented:
Thank you for the review and suggestions! I've updated the PR
1. I simplified the enum incompleteness check using `semanticRun < PASS.semanticdone` as suggested. 

2. And reduced the test file to only include the original case from the issue. and that makes sense since the test suite is already so big, and keeps getting bigger.

dkorpel commented on compiler/src/dmd/statementsem.d:
```diff
@@ -1959,8 +1959,33 @@ Statement statementSemanticVisit(Statement s, Scope* sc)
+            if (ed)
+            {
+                // Check if inside the initializer of one of the enum's members
+                for (Scope* scx = sc; scx; scx = scx.enclosing)
+                {
+                    if (scx.scopesym && scx.scopesym.isEnumDeclaration() && scx.scopesym == ed)
+                    {
+                        error(ss.loc, "cannot use `final switch` on enum `%s` while it is being defined", ed.toChars());
+                        sc.pop();
+                        return setError();
+                    }
+                }
```
This check still needs to be removed

dkorpel commented on compiler/src/dmd/statementsem.d:
```diff
@@ -1959,8 +1959,33 @@ Statement statementSemanticVisit(Statement s, Scope* sc)
                 ed = ds.isEnumDeclaration();
+
+            // Circular references
+            if (ed)
```
```suggestion
            if (ed && ed.semanticRun < PASS.semanticdone)
```

dkorpel commented on compiler/src/dmd/statementsem.d:
```diff
@@ -1959,8 +1959,33 @@ Statement statementSemanticVisit(Statement s, Scope* sc)
+                if (ed.semanticRun < PASS.semanticdone)
+                {
+                    error(ss.loc, "cannot use `final switch` on enum `%s` while it is being defined", ed.toChars());
+                    sc.pop();
```
Why did you add this line?

abulgit replied:
I kept both checks thinking they might catch different cases.

dkorpel replied:
Nevermind this comment, I was asking about `sc.pop()` specifically, but I see now that a scope is pushed before and every return path needs to pop it.

dkorpel commented on compiler/src/dmd/statementsem.d:
```diff
@@ -1959,8 +1959,33 @@ Statement statementSemanticVisit(Statement s, Scope* sc)
+
             if (ed && ss.cases.length < ed.members.length)
             {
+                // Remove the old check for incomplete enum members
```
This looks like an AI instruction accidentally left in here

abulgit replied:
no! I remove the previous enum members checks that's why I added that line.

dkorpel replied:
But the comment doesn't make sense without the context of this Pull Request, and only adds confusion when someone reads it later.

abulgit replied:
```
if (!ed && te && ((ds = te.toDsymbol(sc)) !is null))
                ed = ds.isEnumDeclaration();
            
            if (ed && ed.semanticRun < PASS.semanticdone)
            {
                error(ss.loc, "cannot use `final switch` on enum `%s` while it is being defined", ed.toChars());
                sc.pop();
                return setError();
            }
            
            if (ed && ss.cases.length < ed.members.length)
            {
```
did u mean this,  remove all the other checks only `ed && ed.semanticRun < PASS.semanticdone`

abulgit replied:
> But the comment doesn't make sense without the context of this Pull Request, and only adds confusion when someone reads it later.

I wrote it for my own, mistakenlyforgot to remove it.

dkorpel replied:
> did u mean this

Yes

abulgit replied:
> > did u mean this
> 
> Yes

Yes I got ur point, but the case is only this check is not Sufficient, As I can see now the test case failed with caught signal: -1073741819. 
I think we also need to consider the Scope checking,  for catching in-progress enums. like this- 

```
  if (ed)
  {
      // Check if inside the initializer of the enum's members (Via Scope Chain)
      for (Scope* scx = sc; scx; scx = scx.enclosing)
      {
          if (scx.scopesym && scx.scopesym.isEnumDeclaration() && scx.scopesym == ed)
          {
              error(ss.loc, "cannot use `final switch` on enum `%s` while it is being defined", ed.toChars());
              sc.pop();
              return setError();
          }
      }

      // Check if enum semantic analysis is not yet complete
      if (ed.semanticRun < PASS.semanticdone)
      {
          error(ss.loc, "cannot use `final switch` on enum `%s` while it is being defined", ed.toChars());
          sc.pop();
          return setError();
      }
  }
```

abulgit replied:
or may be we can do something better like Combines both checks into one isInvalid flag -

 ```
 if (ed)
  {
      bool isInvalid = false;

      // Check if inside enum member initializer via scope chain
      for (Scope* scx = sc; scx && !isInvalid; scx = scx.enclosing)
      {
          isInvalid = (scx.scopesym && scx.scopesym.isEnumDeclaration() && scx.scopesym == ed);
      }
      // Combine both checks
      isInvalid = isInvalid || (ed.semanticRun < PASS.semanticdone);

      if (isInvalid)
      {
          error(ss.loc, "cannot use `final switch` on enum `%s` while it is being defined", ed.toChars());
          sc.pop();
          return setError();
      }
  }
```

abulgit replied:
Both pass the testcase.

abulgit replied:
I would love to hear ur suggestion.

dkorpel replied:
Checking in-progress enum declarations should be done with the semanticRun state. If that's wrong, that should be investigated. It should not be covered up by an improvised scope check, because that will make the bug return and make the next bug fix even more complex.

Concretely, the reason the scope check is insufficient is because a function inside the enum scope is just one possible case, while circular references can occur outside the enum as well:
```D
enum E
{
    a = 3,
    b = f(),
}

int f()() 
{
    E e;
    final switch (e)
    {
        case E.a: break;
    }
    return 4;
}
```

abulgit replied:
Okay, I understand your point that the in-progress enum declarations should be done with the semanticRun state. So my question is - 
You mentioned the semanticRun check should be sufficient. but how the enum's semantic state works.
Also, could you suggest specific debugging or maybe some instrumentation I could add to better understand this problem
Because I wanna understand this whole thing.

dkorpel replied:
You could print the value of ed.semanticRun at the time of the crash, and then locate where it was last set with a debugger / debug print statements.

abulgit replied:
I tried adding debug points and print statements to trace how semanticRun interacts with enum member initializers, but things are getting so overwhelming for me.

If you or someone else solve this, I’d love to see the final implementation to better understand how this compiler tracks enum in these edge cases. A concrete example would help me learn how to handle similar issues in the future.
Thanks for your guidance!

abulgit replied:
Yeah, I will give it a try.

abulgit committed:
moved the enum member analysis

abulgit commented:
I don't understand why this failed
```
src/core/time.d(360): Error: cannot use `final switch` on enum `ClockType` while it is being defined
            with(ClockType) final switch (clockType)
                            ^
src/core/time.d(2055): Error: template instance `core.time.MonoTimeImpl!(ClockType.normal)` error instantiating
alias MonoTime = MonoTimeImpl!(ClockType.normal);
                 ^
src/core/time.d(360): Error: cannot use `final switch` on enum `ClockType` while it is being defined
            with(ClockType) final switch (clockType)
                            ^
src/core/time.d(2055): Error: template instance `core.time.MonoTimeImpl!(ClockType.normal)` error instantiating
alias MonoTime = MonoTimeImpl!(ClockType.normal);
                 ^
make: *** [Makefile:420: ../generated/linux/release/64/libdruntime.a] Error 1
make: *** Waiting for unfinished jobs....
make: *** [Makefile:414: ../generated/linux/release/64/libdruntime.so.a] Error 1
make: Leaving directory '/__w/dmd/dmd/druntime'
Error: Process completed with exit code 2.
```

dkorpel commented:
Yes, that's why solving semantic analysis bugs is so hard! In this case, I think semantic2 doesn't get run. It's best to take small steps, so I'd start by simply setting the semantic2done flag at the end of the current enumSemantic function and see if that breaks anything. 

I'd only really extract / move the code as a follow-up refactoring PR.

abulgit committed:
Move enum number analysis to sementic2

abulgit committed:
WhiteSpace Remove

abulgit commented:
> Yes, that's why solving semantic analysis bugs is so hard! In this case, I think semantic2 doesn't get run. It's best to take small steps, so I'd start by simply setting the semantic2done flag at the end of the current enumSemantic function and see if that breaks anything.
> 
> I'd only really extract / move the code as a follow-up refactoring PR.

Yes, I Debugged the code and Tried so many different approaches, It took so many hrs and finally, this worked. 
can u please review this?

abulgit commented on compiler/src/dmd/enumsem.d:
```diff
@@ -718,3 +722,46 @@ void enumMemberSemantic(Scope* sc, EnumMember em)
+ */
+void enumSemantic2(Scope* sc, EnumDeclaration ed)
+{
+    if (ed.semanticRun >= PASS.semantic2done)
+        return;
+
+    if (ed.semanticRun < PASS.semanticdone)
+    {
+        error(ed.loc, "forward reference to enum `%s`", ed.toChars());
+        ed.semanticRun = PASS.semantic2done;
+        return;
+    }
+
+    assert(ed.semanticRun == PASS.semanticdone || ed.semanticRun == PASS.semantic2);
+    ed.semanticRun = PASS.semantic2;
+
+    if (!ed.members || !ed.symtab || ed.errors)
+    {
+        ed.semanticRun = PASS.semantic2done;
+        return;
+    }
+
+    if (sc)
+    {
+        sc = sc.startCTFE();
+        sc.setNoFree();
+    }
+
+    // Ensure all enum members have their values computed
+    foreach (s; *ed.members)
+    {
+        if (EnumMember em = s.isEnumMember())
+        {
+            if (em.semanticRun < PASS.semanticdone)
+                em.dsymbolSemantic(sc);
+        }
+    }
+
+    ed.semanticRun = PASS.semantic2done;
+}
```
Ohh! Yeah I mean ur right. I was following the pattern used for other declarations that have separate semantic2 phases. But in this case, it can be completed entirely in enumSemantic. So the `enumSemantic2` function is unnecessary, I'll remove the redundant function.

abulgit committed:
Remove Redundant Code in enumsem

dkorpel commented:
Well done finishing this!

abulgit commented:
> Well done finishing this!

Thanks, Dennis. I gained a better understanding of the internal workings of this compiler's codebase. Thanks for ur guidance.

ibuclaw commented:
Shouldn't ice fixes have gone to stable?

dkorpel commented:
It's hard to judge, last time it was possibly an ICE fix that broke someone's code between 2.110-rc and 2.110 final.

dkorpel commented:
@abulgit We need to set `ed.semanticRun = PASS.semantic2done;` in `dmd.importc.cEnumSemantic` as well. Do you want to fix this issue?

abulgit commented:
@dkorpel Definitely, I would love to fix that New Issue. 
So, we also need to set `ed.semanticRun = PASS.semantic2done;` at the end of the C enum processing in `enumsem.d`?
